cmd/release-controller/upgrades.go

```diff
@@ -345,3 +349,96 @@ func syncGraphToSecret(graph *UpgradeGraph, update bool, secretClient kv1core.Se
 		glog.V(2).Infof("Saved upgrade graph state to %s/%s", ns, name)
 	}, 5*time.Minute, stopCh)
 }
+
+func releaseTagTimestamp(v string) (time.Time, error) {
+	var t time.Time
+	version, err := semverParseTolerant(v)
+	if err != nil {
+		return t, err
+	}
+	if len(version.Pre) == 0 {
+		return t, fmt.Errorf("tag %s has no timestamp", v)
+	}
+	prereleaseLast := version.Pre[len(version.Pre) - 1].VersionStr
+	releaseTagTimestampFormat := "2006-01-02-150405"
+	if len(prereleaseLast) < len(releaseTagTimestampFormat) {
+		return t, fmt.Errorf("tag %s has no timestamp", v)
+	}
+	return time.Parse(releaseTagTimestampFormat, prereleaseLast[len(prereleaseLast) - len(releaseTagTimestampFormat):])
+}
+
+func (g *UpgradeGraph) Remove(fromTag, toTag string) {
+	if len(fromTag) == 0 || len(toTag) == 0 {
+		return
+	}
+
+	g.lock.Lock()
+	defer g.lock.Unlock()
+	g.removeWithLock(fromTag, toTag)
+}
+
+func (g *UpgradeGraph) removeWithLock(fromTag, toTag string) {
```

Comment:
We generally don't do methods like this unless you're trying to hold the lock the entire time.  When you set up this method, I expected to see PruneGraph take the lock and then invoke `removeWithLock` in a loop, not call `Remove` in a loop.